mediaSession.setCameraActive/setMicrophoneActive callback is executed after WebPage is notified of muted state

youennf opened the pull request:
#### 41195127df9638b14331dc90d5baed7a2d4ea636
<pre>
mediaSession.setCameraActive/setMicrophoneActive callback is executed after WebPage is notified of muted state
<a href="https://bugs.webkit.org/show_bug.cgi?id=279887">https://bugs.webkit.org/show_bug.cgi?id=279887</a>
<a href="https://rdar.apple.com/problem/136219468">rdar://problem/136219468</a>

Reviewed by Eric Carlson.

Let&apos;s start with the following scenario:
1. User starts capturing camera and microphone
2. User mutes capture using Safari UI
3. User tries to unmute capture using MediaSession setCameraActive/setMicrophoneActive

At step 2, WebPage is notified of being muted and triggers MediaSession actions as well as mute/unmute MediaStreamTrack events.
This is all good.

At step 3, WebPageProxy will make a request to grant capture restart.
When this request is granted, WebPageProxy will first mute other pages and unmute itself.
This will trigger new MediaSession actions as well as mute/unmute MediaStreamTrack events.

It will then notify WebPage that the unmute capture request is validated.
This will trigger resolution of the MediaSession setCameraActive/setMicrophoneActive promise.

The consequence is that the order of events/actions and promise resolution is wrong.
In addition, a number of action handlers are executed even though they were not needed since unmuting was requested via setCameraActive/setMicrophoneActive.

To fix this, we add a new callback to UserMediaPermissionRequestProxy which will be called if the request is allowed but before WebPageProxy unmutes itself.
This will give a correct ordering and will remove the additional action handler calls.

Covered by added API test.

* Source/WebKit/UIProcess/UserMediaPermissionRequestProxy.h:
(WebKit::UserMediaPermissionRequestProxy::setBeforeStartingCaptureCallback):
(WebKit::UserMediaPermissionRequestProxy::Function&lt;void):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::willStartCapture):
(WebKit::WebPageProxy::validateCaptureStateUpdate):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKit/GetUserMedia.mm:
(TestWebKitAPI::(WebKit2, ToggleCaptureWhenRestarting)):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/media-session-capture.html: Added.

Canonical link: <a href="https://commits.webkit.org/283925@main">https://commits.webkit.org/283925@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fefacb5adef0a2de4806955a13828d54603a3981

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/67826 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/47208 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/20465 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/71881 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/18967 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/69944 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/55007 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/18773 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/54259 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/12676 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/70893 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/43273 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/58652 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/34726 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/39948 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/16062 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/17325 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/61912 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/16405 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/73579 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/11789 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/15685 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/61712 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/11824 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/58726 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/61738 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15059 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/9610 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/3232 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/43015 "Built successfully") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/44091 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/45278 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/43830 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on Tools/TestWebKitAPI/Tests/WebKit/GetUserMedia.mm:
```diff
@@ -1635,6 +1635,69 @@ function callGetUserMedia(audio, video, successMessage, failureMessage)
+    cameraCaptureStateChange = false;
+    microphoneCaptureStateChange = false;
+
+    [webView setCameraCaptureState:WKMediaCaptureStateMuted completionHandler:nil];
+    [webView setMicrophoneCaptureState:WKMediaCaptureStateMuted completionHandler:nil];
```
It might be useful to pass trivial completion handlers to both of these to also verify that the completion handlers are called at the correct time.

youennf replied:
We do not really tell when they will complete, it is just a completion handler given to SetMuted async IPC.
Maybe we actually should wait for the GPU process to get the mute/unmute request to be processed.
This seems like a good follow-up idea, I'll file a bug.

youennf replied:
I filed https://bugs.webkit.org/show_bug.cgi?id=279986

eric-carlson commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/media-session-capture.html:
```diff
@@ -0,0 +1,104 @@
+    window.actionState += state + ", ";
+}
+
+function getActionState(state)
```
`getActionState` implies that it returns the state, something like `validateActionState` would make more sense.

youennf replied:
Right, wil fix.

youennf committed:
mediaSession.setCameraActive/setMicrophoneActive callback is executed after WebPage is notified of muted state
https://bugs.webkit.org/show_bug.cgi?id=279887
rdar://problem/136219468

Reviewed by Eric Carlson.

Let's start with the following scenario:
1. User starts capturing camera and microphone
2. User mutes capture using Safari UI
3. User tries to unmute capture using MediaSession setCameraActive/setMicrophoneActive

At step 2, WebPage is notified of being muted and triggers MediaSession actions as well as mute/unmute MediaStreamTrack events.
This is all good.

At step 3, WebPageProxy will make a request to grant capture restart.
When this request is granted, WebPageProxy will first mute other pages and unmute itself.
This will trigger new MediaSession actions as well as mute/unmute MediaStreamTrack events.

It will then notify WebPage that the unmute capture request is validated.
This will trigger resolution of the MediaSession setCameraActive/setMicrophoneActive promise.

The consequence is that the order of events/actions and promise resolution is wrong.
In addition, a number of action handlers are executed even though they were not needed since unmuting was requested via setCameraActive/setMicrophoneActive.

To fix this, we add a new callback to UserMediaPermissionRequestProxy which will be called if the request is allowed but before WebPageProxy unmutes itself.
This will give a correct ordering and will remove the additional action handler calls.

Covered by added API test.

* Source/WebKit/UIProcess/UserMediaPermissionRequestProxy.h:
(WebKit::UserMediaPermissionRequestProxy::setBeforeStartingCaptureCallback):
(WebKit::UserMediaPermissionRequestProxy::Function<void):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::willStartCapture):
(WebKit::WebPageProxy::validateCaptureStateUpdate):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKit/GetUserMedia.mm:
(TestWebKitAPI::(WebKit2, ToggleCaptureWhenRestarting)):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/media-session-capture.html: Added.

Canonical link: https://commits.webkit.org/283925@main
